### ISSUE TITLE
Add a way to log when classes are implicitly autowired

### ISSUE THREAD

**Issue by Firehed** [PHP-DI/PHP-DI#608]
Right now, the [documentation section](http://php-di.org/doc/performances.html#optimizing-for-compilation) on optimizing for container compilation makes it clear that for best performance, you want to explicitly list all of the classes to be autowired.

From what I can tell, there's no way to discover that you may have missed one - it will still autowire just fine at runtime even from a cached container, but do so in a slower way.

I think it would be really beneficial to be able to opt-in to logging when this implicit autowiring happens, either to a self-aware `Psr\Log\LoggerInterface` defined in the container or by one explicitly provided.

**Comment by mnapoli** [comment]
This is a good idea! I'm not sure how it would affect performances though 🤔 (i.e. not slow down the container because of that added logic) But it may be worth a try

**Comment by Firehed** [comment]
That's a good point.

At least inside of the CompiledContainer, I don't think there should be any real impact since there's already an existing check to see if the key is defined. Doing it in the base container is probably more complex to not impact performance, but it's also (in my opinion) less beneficial to do there as well.

I could put together a fairly simple patch to add the functionality to the CompiledContainer (only). Depending on how people end up using it, that may be all that's needed.

**Comment by Anyqax** [comment]
an alternative would be to have an option like "autoDiscovery" which when disabled in conjunction with enabled container compilation and the "useAutowiring" flag simply skips runtime auto-discovery of services and crashes on non-configured services?
just logging makes it likely that inexperienced devs never explicitly configure a service for auto-wiring and then  someone has to regularly clean up after them ;)

**Comment by Chi-teck** [comment]
I was about to file the same issue. It seems natural that explicit autowiring through `\DI\autowire()`  does not depend on global `useAutowiring` option.

**Comment by Chi-teck** [comment]
Here is a one-line patch that resolves the issue for me. It can be applied with [composer-patches](https://github.com/cweagans/composer-patches).

[PHP-DI-explicit-autowiring.zip](https://github.com/PHP-DI/PHP-DI/files/5780780/PHP-DI-explicit-autowiring.zip)

**Comment by iliayatsenko** [comment]
@mnapoli, could we merge the patch proposed by @Chi-teck? Because the problem is actual.

**Comment by mnapoli** [comment]
@iliayatsenko I haven't reviewed the patch because it isn't a pull request. Feel free to submit a pull request to get this discussion moving again.

**Comment by iliayatsenko** [comment]
@Firehed, @mnapoli, I solved this problem by implementing following logic (pseudocode).

In 'di_compile.php' command:
```php
  @unlink($compiledContainerFile . $appEnv);
  $builder->addDefinitions(...$definitions);
  $builder->useAutowiring(true); // enable autowiring during compilation, i.e. can use autowire() in definitions.
  $builder->enableCompilation($compiledContainerFile . $appEnv); 
  $builder->build(); // here file is written
```

In application bootstrap:
```php
if ($appEnv->isDev) {
    // in dev env container is not compiled, so add definitions on each request.
    $builder->addDefinitions(...$definitions);
} else {
    // in all other envs we can rely on already compiled container, so no need to add definitions and can disable autowiring.
    $builder->useAutowiring(false);
    $builder->enableCompilation($compiledContainerFile . $appEnv);
}
$builder->build();
``` 

Given that you run di_compile.php command in advance during deployment process and before tests (which run in separated env, not in development), this approach works.

**Comment by alexandrmazur96** [comment]
@iliayatsenko interesting solution. How do you handle situations like "class from the vendor used for autowire but missing in definitions"? E.g.:

```php
use Guzzle\Client as HttpClient;

class Service {
     public function __construct(private HttpClient $client) {}
}
```

The `useAutowiring(false)` here may have a significant impact on the system in production.

------

Pre-compile all vendor packages seems difficult to achieve (need to parse the files, check that class is instantiable and if there were two classes with the same name (stubs in some Symfony packages, for example) - you f*cked), and even if this can be done - it's unlikely you want to have a LOT of garbage inside container hash-table (unnecessary memory consumption, etc).

Without autowiring - a big chance you may get errors in production, with autowiring - you may get a slowdown instantiation approach without even a chance to notice that - it really looks like a hard-detectable place for optimization that has a big performance impact in some cases.

The cherry on the cake - you hardly detect that moment during pull request review, and teammates often forget about such things...

-------

As for me - adding a logger to Container for that case is must have.

**Comment by iliayatsenko** [comment]
@alexandrmazur96, I do not see any difference between vendor and not-vendor classes in this regard. 

My approach implies using autowiring only in 2 cases: 
- during development;
- during deployment, when building pre-compiled container.

In production I turn autowiring off, and only rely on pre-compiled container, created on deployment stage. If some service is absent in pre-compiled container, it will simply cause an error, instead of trying to autowire it "on the fly". This is why I do not need any additional loggers.

Could you please clarify what exact problems do you see in this approach?

**Comment by alexandrmazur96** [comment]
> Could you please clarify what exact problems do you see in this approach?

This :)

> it will simply cause an error

I don't know how fast you guys bring fixes to prod, but it's at least 5-7 minutes for us (without allocating an error and making a branch, writing code, creating PR, and reviewing - just a regular deployment process).

> I do not see any difference between vendor and not-vendor classes

Yeah, I needed to be clearer, sorry. There is no difference between vendor and not-vendor in that case. I thought I could solve the case just by pre-compiling the container cache with all known classes (including vendor) but it's a dead-end, as I mentioned before. And because we can't afford cause an error - we definitely need logger here.

**Comment by alexandrmazur96** [comment]
> all used classes will be collected automatically.

How so? Aren't you required to add a class to the definition list like `SomeClass::class => autowire()` before compiling the container?

> Also, you will always spot this error during tests

This is correct only if you have these tests :)

**Comment by iliayatsenko** [comment]
>How so? Aren't you required to add a class to the definition list like SomeClass::class => autowire() before compiling the container?
Yes, you're right, I still need to do it. I meant that you can list only "top-level" services (for example, controllers), and all the services used by them will be resolved and included in compiled container automatically.

> This is correct only if you have these tests :)
Anyway, I suppose you run your code at least once before deploying it on production : )

**Comment by alexandrmazur96** [comment]
> all the services used by them will be resolved and included in compiled container automatically.

Awesome, I didn't know that, thx!

> Anyway, I suppose you run your code at least once before deploying it on production : )

I do, usually :)

**Comment by iliayatsenko** [comment]
@alexandrmazur96, sorry, I was wrong in this exact statement:
>all the services used by them will be resolved and included in compiled container automatically.

As stated in the docs, you need to explicitly list all the classes you want to appear in the compiled container - https://php-di.org/doc/performances.html#optimizing-for-compilation. So nothing is collected automatically, you're right.

And my approach, with disabling autowiring in favour of pre-compiled container, allows to avoid situations when not-explicitly-listed service is created using reflection in production. Still this approach has a drawback - it requires pre-compilation and testing stage before deployment.

**Comment by alexandrmazur96** [comment]
@mnapoli, if I create a PR with opt-in logging in case of autowiring - is there any chance you'll merge it?

If yes - do you have an idea, what is the best approach to implement it?

**Comment by mnapoli** [comment]
Hi @alexandrmazur96, I'd be ok to merge if the impact to the code is minimal (i.e. I don't want to maintain extra complexity tbh).

Solution 2 sounds better than 1 because it is in line with how other settings are configured. Is there any reason to consider this setting differently?
